### PR TITLE
Source Zuora: fix decimals in stringified datetimes

### PR DESCRIPTION
**Pull request by frans-k**
## What
It outputs timestamps in a format that Zuora can cast to TIMESTAMP.

Today Zuora accepts 6 decimals for fractions of a second when casting to TIMESTAMP, but if that's in a WHERE filter on for example `updateddate`, that filter is ignored for input rows.

So if the table has > 10m rows, it hits the Data Query 10m input row limit. Haven't tested if it affects output rows.

## How
Changed from 6 decimals to 3 for second fractions. Instead of `strftime` it uses Pendulum's `format`.

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

**Comment by sh4sh** [comment]
The tests came up with this error:

```
BUILD FAILED in 22s
> Task :airbyte-integrations:connectors:source-zuora:flakeCheck FAILED
[python] .venv/bin/python -m pflake8 --config /actions-runner/_work/airbyte/airbyte/pyproject.toml ./
	 ./unit_tests/unit_test.py:6:1: F401 'datetime.datetime' imported but unused
```

@frans-k before I remove `datetime.datetime` I want to double check with you since it was added in this PR, is it supposed to be in use somewhere? if not, removing it should get us past the error

File: airbyte-integrations/connectors/source-zuora/unit_tests/unit_test.py [hunk 1]
```diff
@@ -2,7 +2,22 @@
 # Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 #
 
+import unittest
+from datetime import datetime
```
Comment (sh4sh) [inline airbyte-integrations/connectors/source-zuora/unit_tests/unit_test.py:6]:
```suggestion
```

Comment (sh4sh) [reply airbyte-integrations/connectors/source-zuora/unit_tests/unit_test.py:6]: Removing as it's not used

**Comment by sh4sh** [comment]
It looks like the Zuora integration test credentials need to be refreshed. I've submitted a request to get it investigated.

**Comment by frans-k** [comment]
Just let me know if there's anything I can do, @sh4sh.

**Comment by frans-k** [comment]
Good call @jetvp, fixed the OAuth issue.

Now I'm getting a new issue though. Zuora gets spammed with hundreds of `DESCRIBE` queries for the single stream I synced. Zuora only keeps a history of 100 queries, so I don't know how many it's done nor how many it could run if I let it.

Posting a short excerpt from the logs, I can't see anything that stands out however.

```
2023-04-27 14:22:57 INFO i.a.w.i.s.SyncPersistenceImpl(startBackgroundFlushStateTask):168 - starting state flush thread for connectionId 8eb62d62-5ebc-49f1-bb9a-f0ff2df01994
2023-04-27 14:22:57 INFO i.a.w.g.DefaultReplicationWorker(lambda$readFromDstRunnable$4):317 - State in DefaultReplicationWorker from destination: io.airbyte.protocol.models.AirbyteMessage@366e837a[type=STATE,log=<null>,spec=<null>,connectionStatus=<null>,catalog=<null>,record=<null>,state=io.airbyte.protocol.models.AirbyteStateMessage@37b5274b[type=STREAM,stream=io.airbyte.protocol.models.AirbyteStreamState@325523d9[streamDescriptor=io.airbyte.protocol.models.StreamDescriptor@6f5d2204[name=revenueeventitem,namespace=<null>,additionalProperties={}],streamState={"updateddate":"2023-02-01T02:01:29+01:00"},additionalProperties={}],global=<null>,data={"revenueeventitem":{"updateddate":"2023-02-01T02:01:29+01:00"}},additionalProperties={}],trace=<null>,control=<null>,additionalProperties={}]
2023-04-27 14:23:04 INFO i.a.w.g.DefaultReplicationWorker(lambda$readFromDstRunnable$4):317 - State in DefaultReplicationWorker from destination: io.airbyte.protocol.models.AirbyteMessage@2d6a09c6[type=STATE,log=<null>,spec=<null>,connectionStatus=<null>,catalog=<null>,record=<null>,state=io.airbyte.protocol.models.AirbyteStateMessage@33fbfc5f[type=STREAM,stream=io.airbyte.protocol.models.AirbyteStreamState@28ab5a12[streamDescriptor=io.airbyte.protocol.models.StreamDescriptor@2e601001[name=revenueeventitem,namespace=<null>,additionalProperties={}],streamState={"updateddate":"2023-02-01T02:01:29+01:00"},additionalProperties={}],global=<null>,data={"revenueeventitem":{"updateddate":"2023-02-01T02:01:29+01:00"}},additionalProperties={}],trace=<null>,control=<null>,additionalProperties={}]
2023-04-27 14:23:04 WARN i.a.w.i.b.DefaultSyncStatsTracker(updateDestinationStateStats):195 - The message tracker encountered an issue that prevents committed record counts from being reliably computed. This only impacts metadata and does not indicate a problem with actual sync data.
io.airbyte.workers.internal.book_keeping.StateDeltaTracker$StateDeltaTrackerException: State hash -1215753501 was already committed, likely indicating a state hash collision
	at io.airbyte.workers.internal.book_keeping.StateDeltaTracker.commitStateHash(StateDeltaTracker.java:121) ~[io.airbyte-airbyte-commons-worker-0.43.1.jar:?]
	at io.airbyte.workers.internal.book_keeping.DefaultSyncStatsTracker.updateDestinationStateStats(DefaultSyncStatsTracker.java:192) ~[io.airbyte-airbyte-commons-worker-0.43.1.jar:?]
	at io.airbyte.workers.internal.sync_persistence.SyncPersistenceImpl.updateDestinationStateStats(SyncPersistenceImpl.java:426) ~[io.airbyte-airbyte-commons-worker-0.43.1.jar:?]
	at io.airbyte.workers.internal.book_keeping.AirbyteMessageTracker.handleDestinationEmittedState(AirbyteMessageTracker.java:137) ~[io.airbyte-airbyte-commons-worker-0.43.1.jar:?]
	at io.airbyte.workers.internal.book_keeping.AirbyteMessageTracker.acceptFromDestination(AirbyteMessageTracker.java:103) ~[io.airbyte-airbyte-commons-worker-0.43.1.jar:?]
	at io.airbyte.workers.general.DefaultReplicationWorker.lambda$readFromDstRunnable$4(DefaultReplicationWorker.java:319) ~[io.airbyte-airbyte-commons-worker-0.43.1.jar:?]
	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.lang.Thread.run(Thread.java:1589) ~[?:?]
2023-04-27 14:23:04 INFO i.a.w.g.DefaultReplicationWorker(lambda$readFromDstRunnable$4):317 - State in DefaultReplicationWorker from destination: io.airbyte.protocol.models.AirbyteMessage@53391832[type=STATE,log=<null>,spec=<null>,connectionStatus=<null>,catalog=<null>,record=<null>,state=io.airbyte.protocol.models.AirbyteStateMessage@3fcc1f0e[type=STREAM,stream=io.airbyte.protocol.models.AirbyteStreamState@4dc5a92f[streamDescriptor=io.airbyte.protocol.models.StreamDescriptor@638b588e[name=revenueeventitem,namespace=<null>,additionalProperties={}],streamState={"updateddate":"2023-02-01T02:01:30+01:00"},additionalProperties={}],global=<null>,data={"revenueeventitem":{"updateddate":"2023-02-01T02:01:30+01:00"}},additionalProperties={}],trace=<null>,control=<null>,additionalProperties={}]
2023-04-27 14:23:10 INFO i.a.w.g.DefaultReplicationWorker(lambda$readFromDstRunnable$4):317 - State in DefaultReplicationWorker from destination: io.airbyte.protocol.models.AirbyteMessage@78683445[type=STATE,log=<null>,spec=<null>,connectionStatus=<null>,catalog=<null>,record=<null>,state=io.airbyte.protocol.models.AirbyteStateMessage@27037029[type=STREAM,stream=io.airbyte.protocol.models.AirbyteStreamState@47ef3937[streamDescriptor=io.airbyte.protocol.models.StreamDescriptor@7e19466c[name=revenueeventitem,namespace=<null>,additionalProperties={}],streamState={"updateddate":"2023-02-01T02:01:33+01:00"},additionalProperties={}],global=<null>,data={"revenueeventitem":{"updateddate":"2023-02-01T02:01:33+01:00"}},additionalProperties={}],trace=<null>,control=<null>,additionalProperties={}]
2023-04-27 14:23:10 INFO i.a.w.g.DefaultReplicationWorker(lambda$readFromDstRunnable$4):317 - State in DefaultReplicationWorker from destination: io.airbyte.protocol.models.AirbyteMessage@476a9ff6[type=STATE,log=<null>,spec=<null>,connectionStatus=<null>,catalog=<null>,record=<null>,state=io.airbyte.protocol.models.AirbyteStateMessage@7b543d0a[type=STREAM,stream=io.airbyte.protocol.models.AirbyteStreamState@45cc95d6[streamDescriptor=io.airbyte.protocol.models.StreamDescriptor@71528125[name=revenueeventitem,namespace=<null>,additionalProperties={}],streamState={"updateddate":"2023-02-01T02:01:33+01:00"},additionalProperties={}],global=<null>,data={"revenueeventitem":{"updateddate":"2023-02-01T02:01:33+01:00"}},additionalProperties={}],trace=<null>,control=<null>,additionalProperties={}]
2023-04-27 14:23:18 INFO i.a.w.g.DefaultReplicationWorker(lambda$readFromDstRunnable$4):317 - State in DefaultReplicationWorker from destination: io.airbyte.protocol.models.AirbyteMessage@625435d6[type=STATE,log=<null>,spec=<null>,connectionStatus=<null>,catalog=<null>,record=<null>,state=io.airbyte.protocol.models.AirbyteStateMessage@3832f8fb[type=STREAM,stream=io.airbyte.protocol.models.AirbyteStreamState@19f1b80e[streamDescriptor=io.airbyte.protocol.models.StreamDescriptor@7c4f59bc[name=revenueeventitem,namespace=<null>,additionalProperties={}],streamState={"updateddate":"2023-02-01T02:01:33+01:00"},additionalProperties={}],global=<null>,data={"revenueeventitem":{"updateddate":"2023-02-01T02:01:33+01:00"}},additionalProperties={}],trace=<null>,control=<null>,additionalProperties={}]
2023-04-27 14:23:18 INFO i.a.w.g.DefaultReplicationWorker(lambda$readFromDstRunnable$4):317 - State in DefaultReplicationWorker from destination: io.airbyte.protocol.models.AirbyteMessage@3ff62966[type=STATE,log=<null>,spec=<null>,connectionStatus=<null>,catalog=<null>,record=<null>,state=io.airbyte.protocol.models.AirbyteStateMessage@10acfb43[type=STREAM,stream=io.airbyte.protocol.models.AirbyteStreamState@4891090[streamDescriptor=io.airbyte.protocol.models.StreamDescriptor@20d3b04[name=revenueeventitem,namespace=<null>,additionalProperties={}],streamState={"updateddate":"2023-02-01T02:01:33+01:00"},additionalProperties={}],global=<null>,data={"revenueeventitem":{"updateddate":"2023-02-01T02:01:33+01:00"}},additionalProperties={}],trace=<null>,control=<null>,additionalProperties={}]
```

Those last messages are repeated, unclear if they're related to the issue. `data={"revenueeventitem":{"updateddate":"2023-02-01T02:07:29+01:00"}}` seem to be incremented ever so slightly with each log message.

**Comment by jetvp** [comment]
The ZuoraObjectsBase > def get_json_schema(self) should have the follow above it:
```
    @lru_cache(maxsize=None)
    def get_json_schema(self) -> Mapping[str, Any]:
        """
        Override get_json_schema CDK method to retrieve the schema information for Zuora Object dynamicaly.
        """
        schema = list(ZuoraDescribeObject(self.name, config=self._config).read_records(sync_mode=None))
        return {"type": "object", "properties": {key: d[key] for d in schema for key in d}}
```

That fixed up the continuous calls to describe.  I'm currently fixing it to make sure it does not return more than the 50k rows, and itinerates from the last date (but can have that in a separate patch).

**Comment by jetvp** [comment]
@frans-k , let me know if that works for you.

**Comment by frans-k** [comment]
@jetvp Yes, it seems to work now, which of course means I'm running into new issues on the Zuora side.

Unrelated to this PR, but just to share the new issue:

```
select *
            from orderaction where
            updateddate >= TIMESTAMP '2023-02-16 00:54:54.000 +01:00' and
            updateddate <= TIMESTAMP '2023-02-20 23:54:54.000 +00:00'
            order by updateddate asc
```

Fails on `dbValue 3 not supported for coreEnumClass CancellationPolicy`. Guessing that #2227 is required to fix this on the client side.

Even a table that don't have this issue is failing on _something_ as well:

```
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/http.py", line 439, in _read_pages
    request, response = self._fetch_next_page(stream_slice, stream_state, next_page_token)
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/http.py", line 462, in _fetch_next_page
    response = self._send_request(request, request_kwargs)
  File "/airbyte/integration_code/source_zuora/source.py", line 452, in _send_request
    status = job_check["data"]["queryStatus"]
KeyError: 'data'
,retryable=<null>,timestamp=1684489330161], io.airbyte.config.FailureReason@34348470[failureOrigin=source,failureType=<null>,internalMessage=Source process exited with non-zero exit code 1,externalMessage=Something went wrong within the source connector,metadata=io.airbyte.config.Metadata@454b1d9c[additionalProperties={attemptNumber=0, jobId=120, connector_command=read}],stacktrace=io.airbyte.workers.internal.exception.SourceException: Source process exited with non-zero exit code 1
	at io.airbyte.workers.general.DefaultReplicationWorker.lambda$readFromSrcAndWriteToDstRunnable$7(DefaultReplicationWorker.java:457)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
,retryable=<null>,timestamp=1684489330327]],commitStateAsap=true]
```

Don't have the time today to try and debug this any further.

**Comment by marcosmarxm** [comment]
Sorry the long delay to get this reviewed and merged. Our team is finishing the August Hackathon review contributions in next two weeks and after we're going to return to the community backlog.

**Comment by sh4sh** [comment]
@frans-k circling back to this PR and hoping to get it merged soon. Since it's been a while, I wanted to confirm - are you still interested in continuing this work? Thanks in advance!

**Comment by marcosmarxm** [comment]
Closed because user didn't return last comment. @frans-k let us know if you want to continue the work here.

**Pull request closed without merging**